File: ProgrammingProjects/python/chemistry/geometry_analyzer.py
import numpy as np

from .molecule import Molecule


class GeometryAnalyzer:
    def __init__(self, molecule: Molecule) -> None:
        self._molecule = molecule

    def distance(self, i: int, j: int) -> float:
        displacement = (
            self._molecule.atom(i).position
            - self._molecule.atom(j).position
        )
        return float(np.linalg.norm(displacement))

    def distances_matrix(self) -> np.ndarray:
        count = self._molecule.atom_count
        return np.array(
            [
                [
                    self.distance(i, j)
                    for j in range(count)
                ]
                for i in range(count)
            ]
        )

    def angle_degrees(
        self,
        i: int,
        j: int,
        k: int,
    ) -> float:
        ji = (
            self._molecule.atom(i).position
            - self._molecule.atom(j).position
        )
        jk = (
            self._molecule.atom(k).position
            - self._molecule.atom(j).position
        )

        ji_norm = np.linalg.norm(ji)
        jk_norm = np.linalg.norm(jk)

        if ji_norm == 0.0 or jk_norm == 0.0:
            raise ValueError(
                "Cannot calculate an angle "
                "with coincident atoms"
            )

        cosine = np.clip(
            np.dot(ji, jk) / (ji_norm * jk_norm),
            -1.0,
            1.0,
        )
        return float(np.degrees(np.arccos(cosine)))

    def out_of_plane_angle_degrees(
        self,
        i: int,
        j: int,
        k: int,
        l: int,
    ) -> float:
        ki = (
            self._molecule.atom(i).position
            - self._molecule.atom(k).position
        )
        kj = (
            self._molecule.atom(j).position
            - self._molecule.atom(k).position
        )
        kl = (
            self._molecule.atom(l).position
            - self._molecule.atom(k).position
        )

        plane_normal = np.cross(kj, kl)
        ki_norm = np.linalg.norm(ki)
        normal_norm = np.linalg.norm(plane_normal)

        if ki_norm == 0.0:
            raise ValueError(
                "Out-of-plane atom coincides "
                "with the central atom"
            )

        if normal_norm < 1e-12:
            return 0.0

        sine = np.clip(
            np.dot(ki, plane_normal)
            / (ki_norm * normal_norm),
            -1.0,
            1.0,
        )
        return float(np.degrees(np.arcsin(sine)))

    def torsional_angle_degrees(
        self,
        i: int,
        j: int,
        k: int,
        l: int,
    ) -> float:
        ij = (
            self._molecule.atom(j).position
            - self._molecule.atom(i).position
        )
        jk = (
            self._molecule.atom(k).position
            - self._molecule.atom(j).position
        )
        kl = (
            self._molecule.atom(l).position
            - self._molecule.atom(k).position
        )

        ijk_normal = np.cross(ij, jk)
        jkl_normal = np.cross(jk, kl)

        ijk_norm = np.linalg.norm(ijk_normal)
        jkl_norm = np.linalg.norm(jkl_normal)

        if ijk_norm < 1e-12 or jkl_norm < 1e-12:
            raise ValueError(
                "Torsional angle is undefined "
                "for collinear atoms"
            )

        cosine = np.clip(
            np.dot(ijk_normal, jkl_normal)
            / (ijk_norm * jkl_norm),
            -1.0,
            1.0,
        )
        angle = float(np.degrees(np.arccos(cosine)))

        sign = np.dot(
            jk,
            np.cross(ijk_normal, jkl_normal),
        )
        if sign < 0.0:
            angle *= -1.0

        return angle

File: ProgrammingProjects/cpp/common/src/geometry_analyzer.cpp
#include <chemistry/geometry_analyzer.h>

#include <chemistry/physical_constants.h>

#include <Eigen/Geometry>

#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace chemistry {

GeometryAnalyzer::GeometryAnalyzer(const Molecule& molecule)
    : molecule_(molecule)
{
}

double GeometryAnalyzer::distance(std::size_t i, std::size_t j) const
{
    return (
        molecule_.atom(i).position -
        molecule_.atom(j).position
    ).norm();
}

double GeometryAnalyzer::angle_degrees(
    std::size_t i,
    std::size_t j,
    std::size_t k
) const
{
    const Eigen::Vector3d ji =
        molecule_.atom(i).position -
        molecule_.atom(j).position;
    const Eigen::Vector3d jk =
        molecule_.atom(k).position -
        molecule_.atom(j).position;

    const double ji_norm = ji.norm();
    const double jk_norm = jk.norm();

    if (ji_norm == 0.0 || jk_norm == 0.0) {
        throw std::runtime_error(
            "Cannot calculate an angle with coincident atoms"
        );
    }

    const double cosine = std::clamp(
        ji.dot(jk) / (ji_norm * jk_norm),
        -1.0,
        1.0
    );

    return std::acos(cosine) * 180.0 / pi;
}

double GeometryAnalyzer::out_of_plane_angle_degrees(
    std::size_t i,
    std::size_t j,
    std::size_t k,
    std::size_t l
) const
{
    const Eigen::Vector3d ki =
        molecule_.atom(i).position -
        molecule_.atom(k).position;
    const Eigen::Vector3d kj =
        molecule_.atom(j).position -
        molecule_.atom(k).position;
    const Eigen::Vector3d kl =
        molecule_.atom(l).position -
        molecule_.atom(k).position;

    const Eigen::Vector3d plane_normal = kj.cross(kl);
    const double ki_norm = ki.norm();
    const double normal_norm = plane_normal.norm();

    if (ki_norm == 0.0) {
        throw std::runtime_error(
            "Out-of-plane atom coincides with the central atom"
        );
    }

    if (normal_norm < 1e-12) {
        return 0.0;
    }

    const double sine = std::clamp(
        ki.dot(plane_normal) / (ki_norm * normal_norm),
        -1.0,
        1.0
    );

    return std::asin(sine) * 180.0 / pi;
}

double GeometryAnalyzer::torsional_angle_degrees(
    std::size_t i,
    std::size_t j,
    std::size_t k,
    std::size_t l
) const
{
    const Eigen::Vector3d ij =
        molecule_.atom(j).position -
        molecule_.atom(i).position;
    const Eigen::Vector3d jk =
        molecule_.atom(k).position -
        molecule_.atom(j).position;
    const Eigen::Vector3d kl =
        molecule_.atom(l).position -
        molecule_.atom(k).position;

    const Eigen::Vector3d ijk_normal = ij.cross(jk);
    const Eigen::Vector3d jkl_normal = jk.cross(kl);

    const double ijk_norm = ijk_normal.norm();
    const double jkl_norm = jkl_normal.norm();

    if (ijk_norm < 1e-12 || jkl_norm < 1e-12) {
        throw std::runtime_error(
            "Torsional angle is undefined for collinear atoms"
        );
    }

    const double cosine = std::clamp(
        ijk_normal.dot(jkl_normal) / (ijk_norm * jkl_norm),
        -1.0,
        1.0
    );

    double angle = std::acos(cosine) * 180.0 / pi;
    const double sign =
        jk.dot(ijk_normal.cross(jkl_normal));

    if (sign < 0.0) {
        angle *= -1.0;
    }

    return angle;
}

} // namespace chemistry
